CoastalResilienceNetwork/GeositeFramework
# Internationalization Scheme

## Overview

This is an attempt to devise a concrete internationalization scheme for the app that will allow both the core and plugin apps to be localizable for different locales. The locale to use will be specified in the instance's `region.json` file. Thus, the locale selection is instance-wide, and not user-set.

To facilitate string replacement, we use the [i18next](http://i18next.com) library, which comes with an [ecosystem of plugins](http://i18next.com/docs/ecosystem/) which serve most of the functionality we need. In addition to the core library, we use the following plugins:

 * [i18next-xhr-backend](https://github.com/i18next/i18next-xhr-backend): This allows us to specify a pattern for finding `.json` files on the server, and the `i18next` instance on the front-end to request it dynamically.
 * [i18next-jquery](https://github.com/i18next/jquery-i18next): Allows the use of internationalization in HTML templates in the form of `data-i18n` attributes. In addition to the attribute, the tags should have a `.i18n` class applied to them.
 * [i18next-sprintf-postProcessor](https://github.com/i18next/i18next-sprintf-postProcessor): Provides a `printf` like syntax, allowing the translation to have parameters in different positions given the language's grammar.

## Usage

Using this library, we can translate strings to other languages in the following ways:

 * In JavaScript, using `i18next.t('source phrase')`: Once the app is loaded and `i18next` has been initialized, it is attached to the `window` object and thus available everywhere. Within JavaScript, we can simply use `i18next.t('source phrase')` to translate strings. This will look up the `source phrase` in the supplied translation file, and return the found value. If the phrase is not found in the translation file, it will return `source phrase` itself.
 For strings with dynamic parameters, we can use the `printf` syntax:

  ```js
  i18next.t('Longitude %f, Latitude %f', lng, lat);
  ```

 And the translation file should have corresponding markers:

  ```json
  {
      "Longitude %f, Latitude %f": "Länge %f, Breite %f"
  }
  ```
 * In HTML, using the `data-i18n` attribute: This allows the specification of internationalization text right within the templates themselves. At the end of the first render, we go through all the marked tags in the entire app and localize them:

  ```js
  $('.i18n').localize();
  ```

 This replaces the contents of those tags with the translated value of the contents of `data-i18n`. If we wish to replace the value of another attribute, instead of the tag itself, then we prefix the string with the attribute's name in brackets. For example:

  ```html
  <a href="#" class="i18n" data-i18n="[title]source phrase">Text</a>
  ```

 Which outputs:

  ```html
  <a href="#" class="i18n" data-i18n="[title]source phrase" title="translated phrase">Text</a>
  ```

 The `.i18n` class is used by the `.localize()` call above to find and replace the value in these tags. It must also be called on the view's `$el` every time the view renders, since the replacements in the HTML template will have reset. An example of this can be seen in the `renderSidebar` method in `Pane.js`.

## Translations

The translations themselves are stored as simple JSON files, each named for the locale it translates to. There are three locations for them:

 * `/locales/{{ locale }}.json`: The translations for the core app.
 * `/{{ plugin_path }}/locales/{{ locale }}.json`: Translations for the plugin
 * `/languages/{{ locale }}.json`: Combined file containing all translations in the app

Translations for English are not technically necessary, since we fallback to the phrase if we don't find the key, and the phrases are already in English. However, for the purpose of creating a guide to be used for translating to other languages, we maintain an English translation file which simply translates each phrase to itself. For example, `/locales/en.json` has entries like:

```json
{
    "Map Legend": "Map Legend",
    "Hide Legend": "Hide Legend",
    "View Only Map 1": "View Only Map 1",
    "View Only Map 2": "View Only Map 2",
    "Full Map": "Full Map",
    
}
```

Plugins must provide their own translations. As an example, we have supplied them for `sample_plugins/identify_point` in its `locales/en.json`:

```json
{
    "Identify Point": "Identify Point",
    "Identify point sample plugin": "Identify point sample plugin",
    "Click any point on the map to display Latitude and Longitude": "Click any point on the map to display Latitude and Longitude",
    "You clicked on latitude %f longitude %f": "You clicked on latitude %f longitude %f"
}
```

When the app starts, we go through each plugin's and add them to a master file. Finally, we add the core translations. In case of collisions, later additions will overwrite earlier ones. This ensures that no one plugin's translation can override the presentation of the entire app.

The combined files are stored in `/languages/{{ locale }}.json`. This folder is ignored by git, but looked for by `i18next` when the app starts up, as configured under `internationalize` in `App.js`.

## Testing Instructions

Open the app and ensure that the tooltips and labels for map controls in the bottom left appear and work correctly. Ensure that the Identify Point plugin works and appears correctly.

## Code Review Instructions

Ensure that the JavaScript libraries are loaded in the correct order. Suggest ideas for making the translation-merge more efficient (`PrepareLanguages` in `Global.asax.cs`).

## Next Steps

This effort designs a strategy for internationalizing the app and the plugins. This strategy, once validated, must be used to internationalize the rest of the app and plugins. For this task, we have #598.

Connects #589
Connects #597
